### ISSUE TITLE
v2 doc todos checklist

### ISSUE THREAD

**Issue by CaselIT** [sqlalchemy/sqlalchemy#7659]
SQLAlchemy has a large volume of documentation that's currently using a mix of mostly v1 and some v2 style queries especially when talking with the orm.

For v2, while there is no plan to remove `query`, most of the docs should adopt the v2 style of doing things.

- [x] session.query -> session.execute/session.scalars
- [x] update Query API docs so that every method refers to background on migrating to the 2.0 pattern, or the corresponding method on Select.  general format/ language to use is:

```
        .. seealso::

            This API is available using :term:`2.0 style` queries as well,
            documented at:

            * :ref:`queryguide_inspection`
```

- [x] Declarative mappings need to use mapped_column()
- [x] update tutorials to introduce mapped_column(), some tips/etc. regarding "what happened to Column()?" etc. describe they are equivalent Not sure to the extent that the docs will push for “annotation only” style, probably more than less.   New dataclasses docs.  The existing “dataclasses” approach with __sa_field_whatever__ it’s called becomes legacy.
- [x] new DeclarativeBase should be used everywhere instead of declarative_base() function
- [ ] update central mypy docs in ORM guide
- [ ] more links in mypy extension to new mypy docs in ORM guide
- [x] document annotation-enabled mapped_column() usage, likely in the ORM tutorial
- [x] remove dead links to removed sections, running the sphinx build shows a lot of warnings to old sections
- [x] Reorganizing mapper config docs further as it’s very hard to follow between declarative/imperative w/ or w/out dataclasses, want to make room for new dataclasses approach.
- [x] update mypy plugin docs with a full deprecation note at the top
- [x] asyncio docs, note with "uses SQLAlchemy’s future mode, which has several potentially breaking changes. " should be reworded or removed for 2.0 docs
- [ ] update the section [Building Query-Enabled Properties](https://docs.sqlalchemy.org/en/14/orm/join_conditions.html#building-query-enabled-properties) tha does not have a direct 1->2 porting, since the standalone `with_parent` requires a relationship
- [ ] update the error section [An alias is being generated automatically for raw clauseelement](https://docs.sqlalchemy.org/en/20/errors.html#error-xaj1)
- [x] is an equivalent for [`Query.only_return_tuples`](https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html#sqlalchemy.orm.Query.only_return_tuples) needed?
     **a:** no, as this was a workaround for the fact that Query is hardcoded to list of entities for a single entity.  the  Result architecture addresses this issue specifically
- [x] is an equivalent for [Query.params](https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html#sqlalchemy.orm.Query.params) needed? Select.params is not comparable, since it does a different thing enirely
    **a:** no, because Query has no execute() method, there was no way to use bindparam() with Query.  With sesion.execute(), you just pass the params
- [x] is an equivalent for [Query.reset_joinpoint](https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html#sqlalchemy.orm.Query.reset_joinpoint) needed?
   **a:** no, because the primary use for "joinpoint" was to support query.join("stringprop").join("anotherstringprop"), where we didn't know what the "left side" was.   join() with a string like that is not part of Select.join(), and it now always has an  explicit indication of the "left side" for relationship joins.  We also have `join_from()` for even more indication of "left side" when needed.    the other use for "joinpoint" was resetting what `filter_by()` would be relative towards.   a code search through openstack etc. showed that nobody was using reset_joinpoint() at all, so Select.filter_by() is now always relative to the last joined entity.  It would be a major change to revise that now and nobody has had any issue with it.
- [x] is an equivalent for [Query.enable_eagerloads](https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html#sqlalchemy.orm.Query.enable_eagerloads) needed?
   **a:** not unless someone comes to us with a usecase for it.  it was mostly used by Query internally for methods that have other approaches in the newer system, and the logic for rendering joined eager loads or not is now mostly automatic based on the context of the query itself.  it really only applies to "joined eager loading" and someone could just as well use a loader option if they have a Select where an unwanted joined eager load is coming out.

The old migration files should not be updated, since there were "current" for an older version of the docs.

Help is very much appreciated in this task, but please check with us for specific sub-tasks as we might be working on some already

**Comment by jvanasco** [comment]
I added an `easy` tag, as volunteers can work on this by section

**Comment by auvipy** [comment]
can you share an example?

**Comment by CaselIT** [comment]
@auvipy I've provided a list (that at the moment contains only query -> execute/scalars) as the area for improval

**Comment by gordthompson** [comment]
I've submitted a small patch at

https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3625

to see if that will get the ball rolling.

**Comment by gordthompson** [comment]
@auvipy 

> can you share an example?

See e246548cd80a2a423c6041ac05ae6521d1c20b46

**Comment by auvipy** [comment]
i will try to comeback with a PR following the ref commit.

**Comment by provinzkraut** [comment]
I've submitted a PR (#7829) where I started working on some of those changes.
Most of them were pretty straightforward, but there were some areas that need a little more attention that simply replacing the querying style.


## Needs textual changes

These are parts of the documentation that specifically talk about `Query` and some of its inner workings, and where the changes in the example code should be accompanied by appropriate changes in the explanations.

- [Query-time SQL expressions as mapped attribute](https://docs.sqlalchemy.org/en/20/orm/mapped_sql_expr.html#query-time-sql-expressions-as-mapped-attributes)
- [Loading Inheritance Hierarchies](https://docs.sqlalchemy.org/en/20//orm/inheritance_loading.html#loading-inheritance-hierarchies)
- [Wildcard and Exclusionary Options with Multiple-Entity Queries](https://docs.sqlalchemy.org/en/20/orm/loading_columns.html#wildcard-and-exclusionary-options-with-multiple-entity-queries)
- [I’ve created a mapping against an Outer Join, and while the query returns rows, no objects are returned. Why not?](https://docs.sqlalchemy.org/en/20/faq/sessions.html#i-ve-created-a-mapping-against-an-outer-join-and-while-the-query-returns-rows-no-objects-are-returned-why-not)
- [Query-time SQL expressions as mapped attributes](https://docs.sqlalchemy.org/en/20/orm/mapped_sql_expr.html#query-time-sql-expressions-as-mapped-attributes)


## Needs some decision-making 

<dl>
  <dt><a href="https://docs.sqlalchemy.org/en/20/orm/join_conditions.html#building-query-enabled-properties">Building Query-Enabled Properties</a></dt>
  <dd>Not sure how this maps to v2 style since <code>with_parent</code> has to be called with <code>prop</code> now. Example code would probably need to be reworked to include an appropriate relationship?</dd>
  <dt><a href="https://docs.sqlalchemy.org/en/20/orm/loading_relationships.html">Relationship Loading Techniques</a></dt>
  <dd>Most of it is not in 2.0 style, but this is specifically mentioned in <a href="https://docs.sqlalchemy.org/en/20/orm/loading_relationships.html#relationship-loading-with-loader-options">this</a> section. If the examples were to be upgraded to 2.0 style, this should probably be removed. I didn't felt comfortable updating the rest of the examples, even though they don't explicitly mention this anymore, as it would leave this page with a mix between 1.x and 2.0 style</dd>
</dl>

I'm happy to contribute the remaining things as well, but since they represent some more substantial changes to the documentation, I didn't feel comfortable doing so without feedback / didn't want to make those decisions.

**Comment by sandangel** [comment]
hi, how can we install this pre-release with `poetry`?

**Comment by CaselIT** [comment]
> hi, how can we install this pre-release with `poetry`?

never used poetry. with pip you just use `pip install git+https://github.com//sqlalchemy/sqlalchemy.git#egg=sqlalchemy`

**Comment by provinzkraut** [comment]
> never used poetry. with pip you just use `pip install git+https://github.com//sqlalchemy/sqlalchemy.git#egg=sqlalchemy`

You can add git dependencies in poetry as well, although it does not support the "project name" notation in URLs. But if you drop the `egg=sqlalchemy` part, it works just fine:

`poetry add git+https://github.com/sqlalchemy/sqlalchemy.git`

**Comment by tgross35** [comment]
@CaselIT are you familiar with blacken-docs? https://github.com/asottile/blacken-docs maintained by the maker of precommit

Might be a nice extra pc hook to add while updating a lot of code examples

**Comment by tgross35** [comment]
Also, there is a [sphinx plugin that adds spellcheck features](https://sphinxcontrib-spelling.readthedocs.io/en/latest/index.html) via `sphinx-build -b spelling` (I notice spelling is a big chunk of comments you make). I had actually added that as a personal helper at some point, and started working on the SQLA-specific dictionary https://github.com/tgross35/sqlalchemy/commit/0a2c8efeacf285ee6090708d457b37e0348b3376

If you'd be open to adding blacken-docs to the pc hooks and/or adding the sphinx spelling plugin, let me know and I'll create a PR. Of course they wouldn't have to be pipeline enforced, but are nice tools to have available. blacken-docs hook is 0 invasiveness if it's marked to run manually, the sphinx spelling fixture is minimally invasive but adds one requirement to the docs builder (doesn't run by default of course but needs to be in `conf.py`)

**Comment by jvanasco** [comment]
I would much prefer a flake8 for inline docs than black at this time.

zope.SqlAlchemy is (was?) using something similar - IIRC it extracted code from docs and ran it as a test.  But, I believe the way our docs scope variables and interact with dbs is incompatible.

Having gone through and edited the docs several times for misc changes, the reason why I’d like flake8 is just to ensure the scoping of names works on each page.  The docs are massive, and when some things have moved around, they have migrated away into other files and might not have the setup the intro for that section anymore.

**Comment by tgross35** [comment]
There might be a way to configure blacken-docs to make it work - I'll play around just to see. Regarding spelling, it needs the  sphinx cache to be built (I think about 3 minutes for me) but it's instant after that. So not a pre-commit hook kind of thing, more of a "available if you need it" or (maybe) CI kind of thing.

@jvanasco [flake8-rst-docstrings](https://github.com/peterjc/flake8-rst-docstrings) is a plugin that does this, but is the specific problem you're referencing when connected code is split into multiple blocks (with text between)? I'm unsure how it handles. That's one of the things I really wish something core python would copy from rust, how it executes all your docstring code examples as tests with asserts (lets you do stuff like hidden code setup lines too).

If there's no strong opposition to the ideas I'll get something together to see how it works.

**Comment by CaselIT** [comment]
We do have some docstring tests in `test/base/test_tutorials.py` but it's a small part of the test only.

> more of a "available if you need it" or (maybe) CI kind of thing.

we could add a gate that checks for spelling errors

not sure about the flake8 plugin, I've never used it. Thoughts @zzzeek @jvanasco ?

**Comment by zzzeek** [comment]
flake8 will check things, that's fine, but I would like black formatting to be applied to code examples also.

**Comment by jvanasco** [comment]
Sorry, traveling this week and it’s hard searching GitHub on a phone.

Iirc, the plug-in zope.SqlAlchemy uses would treat all the text blocks in a single file - across classes, etc - as a single script.  In terms of scoping, that is something I think SqlAlchemy needs — often times an example uses a variable/concept from another file, so it doesn’t make sense if you’re just looking at a single page on the docs.  The drawback to the plug-in is that i think it actually ran the code, which I think is too complex for our docs as so much setup may be required

**Comment by tgross35** [comment]
Alright, started a draft PR here for discussion https://github.com/sqlalchemy/sqlalchemy/pull/8136

**Comment by sqla-tester** [comment]
**Federico Caselli** has proposed a fix for this issue in the **main** branch:

**Improve v2 documentation** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4285

**Comment by Prescriptor** [comment]
On this documentation page (https://docs.sqlalchemy.org/en/20/orm/quickstart.html#declare-models), in line "addresses: Mapped[list["Address"]] = relationship..." annotation "list" gives an error message: 'type' object is not subscriptable.

**Comment by CaselIT** [comment]
You need to be on python 3.19+ to use the lowercase list. Otherwise it should be List

I don't remember if in the documentation we use the uppercase or lowercase collection. Thanks for reporting in any case

**Comment by CaselIT** [comment]
> I don't remember if in the documentation we use the uppercase or lowercase collection

we seem to use both in more or less equal number. Maybe we could standardize on only one form. @zzzeek do you have preferences regarding `List[X]` vs `list[X]` in the docs (same rule would be valid also for other connection types)?

**Comment by zzzeek** [comment]
I think there is no good answer here.  it's way easier for real world programs to use `list[x]`.  But then we will get confused people for several years until 3.8 is EOL.   maybe we should look at other projects like pydantic and sqlmodel to see what they are doing?

**Comment by provinzkraut** [comment]
Just chiming in here because I recently had to make a similar decision: 

Currently the two most popular solutions are:

1. Use the style of the lowest supported Python version (i.e. 3.8 / `typing.List`)
2. Use some form of [tabs](https://sphinx-design.readthedocs.io/en/latest/tabs.html) to show the code for different Python versions

@zzzeek because you mentioned them: 

- Pydantic has tooling in place to automagically generate [tabs](https://facelessuser.github.io/pymdown-extensions/extensions/tabbed/) for their mkdocs based docs (so option 2 but automated)
- SQLModel uses Python 3.8 compatible annotations
- FastAPI explicitly uses manually created tabs for different Python versions

I ended up doing something similar to what Pydantic does and created a small lib to solve this issue for Sphinx based documentations. But I think the approach of using tabs would be a bit challenging because of the custom theme used by SQLAlchemy.

**Comment by CaselIT** [comment]
thanks for chiming in @provinzkraut. Personally I'm -1 on tabs since we would end up with 2x(or more) example to maintain.

Things are somewhat complicated by the fact that the all the typing versions of standard collections are deprecated since 3.9.

I guess the best approach would be to use the typing collection for another couple of years, until 3.8 becomes eol (if it's not dropped before by sqlalchemy)

**Comment by zzzeek** [comment]
there's no technical reason we can't use tabs and there is nothing special about SQLAlchemy's theme.  the main technical issue with tabs is that lots of our docs are run through doctest and that would highly complicate how that would work, so if we were to do that, we'd likely have to automate rewriting the Python code at build time into tabs.

TBH I think that's all overkill and I am not really comfortable adding more "clicking" to the docs, there was for a long time these SQL click things which I think made the docs more effort to get through.

My vote would be to stick with the typing collections for now.

**Comment by CaselIT** [comment]
ok, makes sense. will open a new issue to track this.

**Comment by Prescriptor** [comment]
> I think there is no good answer here. it's way easier for real world programs to use `list[x]`. But then we will get confused people for several years until 3.8 is EOL...

Thanks a lot. List worked. I will stay on Win7 for the next few years:)
